presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleBreakdownHistogram.java

```diff
@@ -0,0 +1,266 @@
+        this.weights = new double[size];
+        this.counts = new long[size];
+        input.readBytes(
+                Slices.wrappedIntArray(indices),
+                size * SizeOf.SIZE_OF_INT);
+        input.readBytes(
+                Slices.wrappedDoubleArray(weights),
+                size * SizeOf.SIZE_OF_DOUBLE);
+        input.readBytes(
+                Slices.wrappedLongArray(counts),
+                size * SizeOf.SIZE_OF_LONG);
+    }
+
+    protected FixedDoubleBreakdownHistogram(FixedDoubleBreakdownHistogram other)
+    {
+        this.bucketCount = other.bucketCount;
+        this.min = other.min;
+        this.max = other.max;
+        verifyParams(bucketCount, min, max);
+        this.indices = new int[other.indices.length];
+        this.weights = new double[other.weights.length];
+        this.counts = new long[other.counts.length];
+        for (int i = 0; i < other.indices.length; ++i) {
+            this.indices[i] = other.indices[i];
+            this.weights[i] = other.weights[i];
+            this.counts[i] = other.counts[i];
+        }
+    }
+
+    public int getBucketCount()
+    {
+        return bucketCount;
+    }
+
+    public double getMin()
+    {
+        return min;
+    }
+
+    public double getMax()
+    {
+        return max;
+    }
+
+    public double getWidth()
+    {
+        return (max - min) / bucketCount;
+    }
+
+    public int getRequiredBytesForSerialization()
+    {
+        return
+                SizeOf.SIZE_OF_INT + // bucketCount
+                        2 * SizeOf.SIZE_OF_DOUBLE + // min, max
+                        SizeOf.SIZE_OF_INT + // size
+                        indices.length * (SizeOf.SIZE_OF_INT + SizeOf.SIZE_OF_DOUBLE + SizeOf.SIZE_OF_LONG);
+    }
+
+    public void serialize(SliceOutput out)
+    {
+        out.appendInt(bucketCount);
+        out.appendDouble(min);
+        out.appendDouble(max);
+        out.appendInt(indices.length);
+        int i = 0;
+        for (i = 0; i < bucketCount; ++i) {
+            out.appendDouble(indices[i]);
+        }
+        for (i = 0; i < bucketCount; ++i) {
+            out.appendDouble(weights[i]);
+        }
+        for (i = 0; i < bucketCount; ++i) {
+            out.appendDouble(counts[i]);
+        }
+    }
+
+    public void add(double value)
+    {
+        add(value, 1.0);
+    }
+
+    public void add(double value, double weight)
+    {
+        add(value, weight, 1);
+    }
+
+    public void add(double value, double weight, long count)
+    {
+        int bucketIndex = getIndexForValue(bucketCount, min, max, value);
+
+        add(bucketIndex, weight, count);
+    }
+
+    private void add(int bucketIndex, double weight, long count)
+    {
+        int index = lowerBound(bucketIndex, weight);
+        if (index < indices.length && indices[index] == bucketIndex && weights[index] == weight) {
+            counts[index] += count;
+            return;
+        }
+
+        int[] newIndices = new int[indices.length + 1];
```

Comment:
Maybe I'm missing something, but it looks like this is not using some sort of exponential growth factor.  Ie. we copy all of the arrays and leave room for just one more element, rather than some sort of growth factor.  Is that correct?  If so, I would consider a scheme like this.

Reply:
Say the length of the stream is $n$ and the eventual size of the histogram is $m$. For the use-case for which this is optimized, $n \ll m$. Then what you're suggesting is $n + \Theta(m) \simeq n$ ops, and with expected size $1.5m$. What's implemented here has $n + \Theta(m^2) \simeq n$ ops, but size $m \lneq 1.5m$.

Reply:
Added implNotes to the class doc.